### ISSUE TITLE
UnobservedTaskException - ObjectDisposedException in MobileAuthenticatedStream

### ISSUE THREAD

**Issue by bocciab** [mono/mono#8350]
In the last week, we've had 59,891 distinct devices encounter it a total of 282,388 times. It appears this affects 4.0 and 4.2.1:

```
[Events]
Source: UnobservedTaskException

[Exception]

System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MobileAuthenticatedStream'.
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.2.1.28/src/Xamarin.Mac/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:152 
  at Mono.Net.Security.MobileAuthenticatedStream+<StartOperation>d__58.MoveNext () [0x001b9] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.2.1.28/src/Xamarin.Mac/mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs:
```
----
```
[Events]
Source: UnobservedTaskException

[Exception]

System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MobileAuthenticatedStream'.
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.0.0.216/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:152 
  at Mono.Net.Security.MobileAuthenticatedStream+<StartOperation>d__58.MoveNext () [0x001b9] in /Library/Frameworks/Xamarin.Mac.framework/Versions/4.0.0.216/src/mono/mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs:380 
```

I don't have a specific repro, because the stack trace is incomplete due to being an unobserved task. We had reported this on https://bugzilla.xamarin.com/show_bug.cgi?id=61002, but that was closed without fully resolving our issue.

This could also be related to the following: 
https://bugzilla.xamarin.com/show_bug.cgi?id=55872 - Similar error, but not in an async/await context
https://github.com/mono/mono/issues/6422 - Similar error, but not in an async/await context
https://github.com/mono/mono/pull/6125 - suggested as a fix for some of the above, but I can't find it in any releases?

I can't find any of those in any current or upcoming release notes, though and I can't confirm they are related beyond having similar errors.

**Comment by brendanzagaeski** [comment]
The merged pull request #6125 does make significant changes to the code paths related to these exceptions. See also https://github.com/mono/mono/issues/7194 for some discussion about some _different_ unobserved exceptions that arose _after_ #6125 and how those were fixed.

The changes from #6125 and the follow-up fixes for #7194 are part of the Mono `2018-02` branch. That Mono branch has not yet been integrated into the [xamarin-macios](https://github.com/xamarin/xamarin-macios) or [xamarin-android](https://github.com/xamarin/xamarin-android) SDKs. The Mono `2018-02` branch would by default tentatively be included in the upcoming (not yet created) `d15-8` branches of the Xamarin SDKs.

As an example of how #6125 changes the code path, the `EndWrite()` method that was involved in the fix for [Xamarin Bugzilla 61002](https://bugzilla.xamarin.com/show_bug.cgi?id=61002) no longer even exists on the `2018-02` branch: <https://github.com/mono/mono/blob/2018-02/mcs/class/System/System.Net/WebConnection.cs>.

Because the `d15-7` Xamarin SDK branches are nearing completion at this time, and since the old code path no longer exists as of the `2018-02` branch, the most likely outcome for this GitHub issue is that it will be closed as resolved (by #6125) on the upcoming `d15-8` Xamarin SDK branches / milestone.

**Comment by bocciab** [comment]
Given the size of impact on our users, is there anything we can do between now and d15-8 to mitigate the issue? What's the ETA for d15-8?

**Comment by brendanzagaeski** [comment]
The first set of published packages based on the `d15-8` branches would be Visual Studio 2017 version 15.8 Preview 1. Based on historical precedent, that version is likely to be released within a week or two of the first Visual Studio 2017 version 15.7 non-preview release. And based on historical precedent, the first version 15.7 non-preview release is itself likely not far away (weeks). See for comparison and reference:

- <https://docs.microsoft.com/visualstudio/install/visual-studio-build-numbers-and-release-dates>
- <https://docs.microsoft.com/visualstudio/productinfo/vs2017-release-rhythm>

* * *

It's probably not relevant, but just in case your application might be accessing network streams via any APIs that don't pass through `WebRequest.SendAsync()` (such as the methods of `NetworkStream` itself), note that it is possible that your application might be missing a call to `EndWrite()`. Other than that, I'm afraid I've exhausted all the creative ideas I had about how an unobserved exception might still happen with the `WebRequest.SendAsync()` code path, so the Mono team would need some clues about what might be special about your application that doesn't happen in other `WebRequest` stress-testing scenarios (such as the scenarios from [Xamarin Bugzilla 61002 , Comment 46](https://bugzilla.xamarin.com/show_bug.cgi?id=61002#c46)).

**Comment by marek-safar** [comment]
I am closing the issue for now as we believe it should be fixed in 15.8 release. 

Please share a repro we could use to verify your use case or re-test with the 15.8 version once you can access it and re-open if you can still reproduce it

**Comment by wilks-ifit** [comment]
How does the versioning that @brendanzagaeski mentioned map to VS for Mac release numbers? I'm getting hundreds of these crashes per day as well.

**Comment by brendanzagaeski** [comment]
The new [Xamarin.Mac version 4.5 Preview](https://developer.xamarin.com/releases/mac/xamarin.mac_4/xamarin.mac_4.5/) and [Xamarin.iOS version 11.11 Preview](https://developer.xamarin.com/releases/ios/xamarin.ios_11/xamarin.ios_11.11/) that have been published today alongside [Visual Studio 2017 for Mac version 7.6 Preview 1](https://docs.microsoft.com/visualstudio/releasenotes/vs2017-mac-preview-relnotes) are based on the Mono `2018-02` branch. They include the [changes to the `HttpWebRequest` async handling](http://www.mono-project.com/docs/about-mono/releases/5.12.0/#class-libraries) from #6125 that will hopefully eliminate any remaining unobserved task exceptions from `MobileAuthenticatedStream` during asynchronous HTTPS requests.

The corresponding update for Xamarin.Android is not yet included in the this first set of preview versions, but it is targeted for inclusion alongside the _next_ Visual Studio 2017 for Mac version 7.6 Preview and the matching Visual Studio 2017 version 15.8 Preview. Similarly, the new Xamarin.Mac and Xamarin.iOS previews are currently only available on Mac. For Windows, they are targeted for inclusion in the _next_ Visual Studio 2017 version 15.8 Preview.

**Comment by brendanzagaeski** [comment]
I'm afraid I don't have more details on timing than that the full release schedule for Visual Studio 2017 version 15.8 should in theory be roughly similar to what the schedule was for Visual Studio 2017 version 15.7. See also the previously linked document on [Visual Studio 2017 build numbers and release dates](https://docs.microsoft.com/visualstudio/install/visual-studio-build-numbers-and-release-dates). One likely reason that Visual Studio 2017 version 15.8 Preview 2 is later compared to the corresponding previews for version 15.7 is due to alignment of the Visual Studio 2017 version 15.8 Preview 1 with the Build conference. Based on the history of the Visual Studio 2017 version 15.7 Previews, it seems likely that Visual Studio 2017 version 15.8 Preview 2 is close on the horizon.

**Comment by bocciab** [comment]
It seems like 15.8 and mono 5.12 are out, but I don't see this bug or bugzilla 61002 mentioned in the release notes. Can anyone confirm the fix for this issue is out?

**Comment by brendanzagaeski** [comment]
Yes, the changes from https://github.com/mono/mono/pull/6125 are included in the current Mono and Xamarin SDK versions (that are part of the Visual Studio 2017 version 15.8 and Visual Studio 2017 for Mac version 7.6 releases). The corresponding section from the Mono release notes is under the [Class Libraries section](https://www.mono-project.com/docs/about-mono/releases/5.12.0/#class-libraries) starting at "`HttpWebRequest` async handling has been rewritten." That's the note for https://github.com/mono/mono/pull/6125, which is the pull request that eliminated the original code path involved in this report.

As one quick extra check to make sure the expected changes were definitely included in the published Xamarin SDK packages, I looked at the IL disassembly of the `System.Net.WebConnection` class from System.dll in each of Xamarin.Mac, Xamarin.iOS, and Xamarin.Android. The `EndWrite()` method correctly no longer exists in that class in the current versions. So the original code path involved in this issue indeed no longer exists.

Note that the `Mono.Net.Security.MobileAuthenticatedStream.StartOperation()` method that appeared in the UnobservedTaskException in the first comment *does* still exist, but the way it is called during an `HttpWebRequest` has changed significantly. If you see an UnobservedTaskException during an `HttpWebRequest` in these new versions, be sure to report back on that. Thanks!

**Comment by softsan** [comment]
After updating VS for Mac to 7.6. i got lots of error during compile time of my existing project which was working fine with the earlier version:
e.g.
if i have a variable like this:
`private string _url;`
it shows following error:
`Predefine type "system.string" is not defined or imported.`

There are lots of similar errors.
So any thoughts on this?

**Comment by wilks-ifit** [comment]
We followed this workaround: https://developercommunity.visualstudio.com/content/problem/317215/after-version-76-all-my-xamarin-forms-project-wont.html#reply-317644 - fixed iOS for me, but Android still has issues

**Comment by softsan** [comment]
Thanks. Yeh correct! that works for PCL and iOS. but Android still having some issue.

**Comment by softsan** [comment]
Upgraded to VS for mac Version 7.6.1 (build 9). please find below details:

**Runtime:**
	Mono 5.12.0.301 (2018-02/4fe3280bba1) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

**Xamarin.Android**
Version: 9.0.0.19 (Visual Studio Professional)

Now i am not able to upload an apk to google play store. 
When i do signing of apk using  Build->Archive option  with the keystore file (Which i was using so far and there was no problem at all! We have 52 release in play store so far with the same keystore and i also double check the keystore validity and its expire in 2042, so that is not an issue). 
It gives me following error while uploading to google play stores:
Upload failed
`You uploaded an APK with an invalid signature (learn more about signing). Error from apksigner: ERROR (Jar signer MYPROJECT.RSA): JAR signature META-INF/MYPROJECT.SF indicates the APK is signed using APK Signature Scheme v2 but no such signature was found. Signature stripped?
`
This was working fine before i upgraded my VS for mac to 7.6.1.
So is there anything i missed or any thoughts? This is currently a blocking issue and i can't publish it to google play.
Thanks.

**Comment by brendanzagaeski** [comment]
> You uploaded an APK with an invalid signature

I believe you are seeing the issue that is being tracked in the following items on Visual Studio Developer Community. I would recommend to vote on and follow those items for updates, potential workarounds, and fix notifications:

- <https://developercommunity.visualstudio.com/content/problem/317799/major-error-76-build-2190-creates-invalid-apk-for.html>
- <https://developercommunity.visualstudio.com/content/problem/308087/apk-signing-succeeds-but-google-yells-about-it.html>

**Comment by michaldobrodenka** [comment]
My application is compiled with latest VS 15.8 (Mono 5.12), beta version of app is compiled with latest Vs 15.9 preview (Mono 5.14) and according to HockeyApp, more than 80% of crashes are Unobserved crash exceptions within MobileAuthenticatedStream. I'm voting for reopening this issue.

```
android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
at caused.AggregateException(s)
Object name: 'MobileAuthenticatedStream'.
at Mono.Net.Security.MobileAuthenticatedStream.<StartOperation>d__58.MoveNext()<ace3a411a33d4fea9c923bdd93cb3ad6>:0
--- End of inner exception stack trace ---
---> (Inner Exception #0) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MobileAuthenticatedStream'.
at Mono.Net.Security.MobileAuthenticatedStream.<StartOperation>d__58.MoveNext()<ace3a411a33d4fea9c923bdd93cb3ad6>:0
```

**Comment by michaldobrodenka** [comment]
Maybe it is related to using websocket-sharp library, which uses Begin/End Read IAsyncResult pattern. These are implemented using tasks, but probably not observed?

**Comment by pcoffer** [comment]
Still happening for me as well with the latest VS 15.8.5.  My stack trace is identical to, and as unhelpful as, that of @michaldobrodenka  above.

**Comment by brendanzagaeski** [comment]
> Maybe it is related to using websocket-sharp library, which uses Begin/End Read IAsyncResult pattern. These are implemented using tasks, but probably not observed?

Appearances of unobserved exceptions from `MobileAuthenticatedStream` could indeed be caused by calls that use `NetworkStream` via some code path other than `HttpClient.SendAsync()`. Any library that interacts with network streams directly rather than via higher level APIs would need to ensure that it called `EndWrite()`. If by chance anyone has a test case they can provide where one of the higher level APIs is leading to unobserved task exceptions, please do open a new issue and attach it so the team can take a look. Thanks!

**Comment by RealDigitApps** [comment]
+1

**Comment by jonathanantoine** [comment]
Hello, 

Why this issue is closed if it still exists on Android ?

**Comment by michaldobrodenka** [comment]
Maybe because it is a duplicate of #2159, it seems fix had been merged already.

**Comment by jonathanantoine** [comment]
Do you know when we should see this in "production nugets" ?
